app/src/main/java/fr/free/nrw/commons/contributions/ContributionsPresenter.java

```diff
@@ -3,104 +3,155 @@
+import fr.free.nrw.commons.di.CommonsApplicationModule;
+import fr.free.nrw.commons.mwapi.UserClient;
+import fr.free.nrw.commons.utils.ExecutorUtils;
+import fr.free.nrw.commons.utils.NetworkUtils;
+import io.reactivex.Observable;
+import io.reactivex.Scheduler;
+import io.reactivex.SingleObserver;
+import io.reactivex.android.schedulers.AndroidSchedulers;
+import io.reactivex.disposables.CompositeDisposable;
+import io.reactivex.disposables.Disposable;
+import io.reactivex.schedulers.Schedulers;
 import timber.log.Timber;
 
-import static fr.free.nrw.commons.contributions.ContributionDao.Table.ALL_FIELDS;
-import static fr.free.nrw.commons.contributions.ContributionsContentProvider.BASE_URI;
-import static fr.free.nrw.commons.settings.Prefs.UPLOADS_SHOWING;
+import static fr.free.nrw.commons.contributions.Contribution.STATE_COMPLETED;
 
 /**
  * The presenter class for Contributions
  */
-public class ContributionsPresenter extends DataSetObserver implements UserActionListener {
+public class ContributionsPresenter implements UserActionListener {
 
     private final ContributionsRepository repository;
+    private final Scheduler mainThreadScheduler;
+    private final Scheduler ioThreadScheduler;
+    private CompositeDisposable compositeDisposable;
     private ContributionsContract.View view;
-    private Cursor cursor;
+    private List<Contribution> contributionList=new ArrayList<>();
 
     @Inject
     Context context;
 
     @Inject
-    ContributionsPresenter(ContributionsRepository repository) {
+    UserClient userClient;
+
+    @Inject
+    AppDatabase appDatabase;
+
+    @Inject
+    SessionManager sessionManager;
+    private LifecycleOwner lifeCycleOwner;
+
+    @Inject
+    ContributionsPresenter(ContributionsRepository repository, @Named(CommonsApplicationModule.MAIN_THREAD) Scheduler mainThreadScheduler,@Named(CommonsApplicationModule.IO_THREAD) Scheduler ioThreadScheduler) {
         this.repository = repository;
+        compositeDisposable=new CompositeDisposable();
+        this.mainThreadScheduler=mainThreadScheduler;
+        this.ioThreadScheduler=ioThreadScheduler;
     }
 
+    private String user;
+
     @Override
     public void onAttachView(ContributionsContract.View view) {
         this.view = view;
-        if (null != cursor) {
-            try {
-                cursor.registerDataSetObserver(this);
-            } catch (IllegalStateException e) {//Cursor might be already registered
-                Timber.d(e);
-            }
-        }
+        compositeDisposable=new CompositeDisposable();
     }
 
-    @Override
-    public void onDetachView() {
-        this.view = null;
-        if (null != cursor) {
-            try {
-                cursor.unregisterDataSetObserver(this);
-            } catch (Exception e) {//Cursor might not be already registered
-                Timber.d(e);
-            }
-        }
+    public void setLifeCycleOwner(LifecycleOwner lifeCycleOwner){
```

Comment:
I would generally just observe liveData in my view.

Also the fragment is passing in the wrong lifecycle, you need to use `getViewLifecycleOwner()`, see [here](https://proandroiddev.com/5-common-mistakes-when-using-architecture-components-403e9899f4cb)

Reply:
> I would generally just observe liveData in my view.
I would have done that too, if it was a view model, as we are following the MVP architecture, I see no harm in making the presenter observe it for the Fragment, if you could elaborate the ill-effects of this @macgills 

> Also the fragment is passing in the wrong lifecycle, you need to use getViewLifecycleOwner(), see here
True, thanks for pointing this out

Reply:
Yeah I just hate MVP.

You could do `Presenter.fetchData`, `View.renderData(LiveData<>)`. It'll leak some logic though. When it comes down to it LiveData is a MVVM component, it is not meant to be used like this.
You could put the logic in the Presenter and return types `Result(List<>)` `NoResults` but it just isn't as cool when it isn't kotlin sealed classes, in java you are reduced to using `instanceOf`. MediatorLiveData exists for this case.